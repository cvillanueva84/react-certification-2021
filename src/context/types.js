// Profile types
export const HOVER_PROFILE = 'HOVER_PROFILE';
export const BLUR_PROFILE = 'BLUR_PROFILE';
// NavBar types
export const TYPE_INPUT = 'TYPE_INPUT';
export const TOGGLE_DARK_MODE = 'TOGGLE_DARK_MODE';
// VideoList types
export const LOAD_VIDEOS = 'LOAD_VIDEOS';
export const SELECT_VIDEO = 'SELECT_VIDEO';
export const TOGGLE_OPENED_DETAILS = 'TOGGLE_OPENED_DETAILS';
