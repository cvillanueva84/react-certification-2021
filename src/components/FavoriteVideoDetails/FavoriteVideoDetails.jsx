import React from 'react';
import { Link, useParams } from 'react-router-dom';

import { useFavorites } from '../../utils/hooks/useFavorites';
import {
  Button,
  Container,
  Grid,
  GridItem,
  List,
  ListItem,
  ListItemAvatar,
  ListItemBody,
  Loader,
  Typography,
} from '../UI';

const FavoriteVideoDetails = () => {
  const { id } = useParams();
  const {
    addFavorite,
    removeFavorite,
    isFavorite,
    favoriteVideos,
    videoSelected,
    isLoading,
  } = useFavorites(id);

  return (
    <Container>
      {isLoading ? (
        <Loader />
      ) : (
        <>
          <Grid container>
            <GridItem xs={12} md={8}>
              <iframe
                src={`https://www.youtube.com/embed/${id}`}
                frameBorder="0"
                title="video"
                width="100%"
                height="600px"
              />
              {isFavorite(id) ? (
                <Button
                  icon
                  color="success"
                  onClick={() => removeFavorite(id)}
                  className="mb-2"
                >
                  <i className="fas fa-star fa-lg" />
                </Button>
              ) : (
                <Button color="info" onClick={() => addFavorite(id)} className="mb-2">
                  <span>Add favorite</span>
                  <i className="fas fa-star fa-lg" />
                </Button>
              )}
              {videoSelected && (
                <>
                  <Typography tagName="h2" className="mb-1" weight="600">
                    {videoSelected.snippet.title}
                  </Typography>
                  <Typography tagName="p">{videoSelected.snippet.description}</Typography>
                </>
              )}
            </GridItem>
            <GridItem xs={12} md={4}>
              <List>
                {favoriteVideos.map((item) => (
                  <Link key={item.id} to={`/favorite/${item.id}`}>
                    <ListItem type="button">
                      <ListItemAvatar
                        src={item.snippet.thumbnails.medium.url}
                        alt={item.snippet.title}
                      />
                      <ListItemBody>
                        <Typography tagName="h5">{item.snippet.title}</Typography>
                        <Typography tagName="p" tiny>
                          {item.snippet.channelTitle}
                        </Typography>
                      </ListItemBody>
                    </ListItem>
                  </Link>
                ))}
              </List>
            </GridItem>
          </Grid>
        </>
      )}
    </Container>
  );
};

export default FavoriteVideoDetails;
