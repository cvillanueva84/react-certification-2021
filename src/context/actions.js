import Types from '../types';

export const contextValueActions = (dispatch) => ({
  search: (payload) => dispatch({ type: Types.SEARCH, payload }),
  theme: () => dispatch({ type: Types.THEME }),
  auth: (payload) => dispatch({ type: Types.AUTH, payload }),
  authLogout: () => dispatch({ type: Types.LOGOUT }),
  videoList: (payload) => dispatch({ type: Types.VIDEO_LIST, payload }),
});
