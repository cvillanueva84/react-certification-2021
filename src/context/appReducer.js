import { types } from '../types/types';

export const appReducer = (state = {}, action) => {
  switch (action.type) {
    case types.search:
      return {
        ...state,
        ...action.payload,
      };
    case types.showVideoDetails:
      return {
        ...state,
        showVideoDetails: true,
      };
    case types.showVideoList:
      return {
        ...state,
        showVideoDetails: false,
      };
    default:
      return state;
  }
};
