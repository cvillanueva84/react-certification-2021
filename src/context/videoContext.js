import React, { createContext, useEffect, useState } from 'react';
import useVideoApi from '../hooks/useVideoApi';

export const VideosContext = createContext(null);

const VideoContext = ({ children }) => {
  const { data, loading, error, fetchVideos } = useVideoApi();
  const [videoSelected, setVideoSelected] = useState(null);
  const [queryVideo, setQueryVideo] = useState('react');

  useEffect(() => {
    getVideos();
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [queryVideo]);

  const getVideos = () => {
    const params = { q: queryVideo };
    fetchVideos({ params });
  };

  const getVideoDetails = (videoID) => {
    const params = { relatedToVideoId: videoID, type: 'video' };
    fetchVideos({ params });
  };

  const selectVideo = (videoDetails) => {
    setVideoSelected(videoDetails);
  };

  return (
    <VideosContext.Provider
      value={{
        data,
        error,
        loading,
        videoSelected,
        getVideos,
        getVideoDetails,
        selectVideo,
        setQueryVideo,
      }}
    >
      {children}
    </VideosContext.Provider>
  );
};

export default VideoContext;
