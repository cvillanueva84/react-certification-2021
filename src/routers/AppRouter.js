import React, { useContext } from 'react';
import { BrowserRouter as Router, Switch } from "react-router-dom";
import { GlobalContext } from '../components/Context';
import { LoginView } from '../components/LoginView';
import { DashboardRoutes } from './DashboardRoutes';
import { PrivateRoute } from './PrivateRoute';
import { PublicRoute } from './PublicRoute';

export const AppRouter = () => {

    const { myStateReducer } = useContext(GlobalContext);

    return (
        <Router>
            <>
                <Switch>
                    <PublicRoute 
                        exact path="/login" 
                        component={LoginView} 
                        isAuthenticated={myStateReducer.logged}
                    />

                    <PrivateRoute 
                        path="/" 
                        component={DashboardRoutes} 
                        isAuthenticated={myStateReducer.logged}
                    />
                </Switch>
            </>
        </Router>
    )
}
