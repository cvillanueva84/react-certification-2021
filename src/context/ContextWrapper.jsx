import React, { useCallback, useReducer } from 'react';
import { AppContext } from './appContext';
import { appReducer } from './appReducer';
import Types from '../types';

const init = {
  search: 'wizeline',
  theme: true,
  isAuthenticated: false,
  auth: {},
  videoList: [],
};

export const ContextWrapper = ({ children }) => {
  const [state, dispatch] = useReducer(appReducer, init);

  const contextValue = {
    state,
    auth: useCallback((payload) => dispatch({ type: Types.AUTH, payload }), []),
    search: useCallback((payload) => dispatch({ type: Types.SEARCH, payload }), []),
    theme: useCallback(() => dispatch({ type: Types.THEME }), []),
    authLogout: useCallback(() => dispatch({ type: Types.LOGOUT }), []),
    videoList: useCallback(
      (payload) => dispatch({ type: Types.VIDEO_LIST, payload }),
      []
    ),
  };

  return <AppContext.Provider value={contextValue}>{children}</AppContext.Provider>;
};
