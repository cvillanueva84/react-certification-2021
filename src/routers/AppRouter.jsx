import React, { useContext, useEffect, useState } from 'react';
import { Switch, Redirect, Route } from 'react-router-dom';

import Favorites from '../components/Favorites';
import FavoriteVideoDetails from '../components/FavoriteVideoDetails';
import Home from '../components/Home/Home.component';
import Page404 from '../components/Page404';
import VideoDetails from '../components/VideoDetails';
import { AppContext } from '../context/appContext';
import { firebase } from '../firebase/firebase.config';
import { PrivateRoute } from './PrivateRoute';

export const AppRouter = () => {
  const { auth } = useContext(AppContext);
  const [isLoggedIn, setIsLoggedIn] = useState(false);

  useEffect(() => {
    firebase.auth().onAuthStateChanged(async (user) => {
      if (user?.uid) {
        auth({
          isAuthenticated: true,
          auth: {
            uid: user.uid,
            email: user.email,
          },
        });
        setIsLoggedIn(true);
      } else {
        setIsLoggedIn(false);
      }
    });
  }, [auth]);

  return (
    <Switch>
      <Route path="/" exact component={Home} />
      <Route path="/video/:id" exact component={VideoDetails} />
      <PrivateRoute
        isAuthenticated={isLoggedIn}
        component={Favorites}
        path="/favorites"
        exact
      />
      <PrivateRoute
        isAuthenticated={isLoggedIn}
        component={FavoriteVideoDetails}
        path="/favorite/:id"
        exact
      />
      <Route path="/404" component={Page404} />
      <Redirect to="/404" />
    </Switch>
  );
};
