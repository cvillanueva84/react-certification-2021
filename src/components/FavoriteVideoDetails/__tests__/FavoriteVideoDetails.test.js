import React from 'react';
import ReactDOM from 'react-dom';
import { render } from '@testing-library/react';
import { HashRouter } from 'react-router-dom';

import FavoriteVideoDetails from '../index';
import { AppRouter } from '../../../routers/AppRouter';
import { ContextWrapper } from '../../../context/ContextWrapper';
import { ThemeProvider } from '../../../ThemeProvider';
// import data from '../../../mocks/youtube.json';

describe('Testing FavoriteVideoDetails component', () => {
  let tree;

  beforeAll(() => {
    ReactDOM.createPortal = jest.fn((element) => {
      return element;
    });
  });

  afterEach(() => {
    ReactDOM.createPortal.mockClear();
  });

  beforeEach(() => {
    tree = render(
      <ContextWrapper>
        <ThemeProvider>
          <HashRouter>
            <AppRouter />
            <FavoriteVideoDetails />
          </HashRouter>
        </ThemeProvider>
      </ContextWrapper>
    );
  });

  it('should match snapshot', () => {
    expect(tree).toMatchSnapshot();
  });
});
