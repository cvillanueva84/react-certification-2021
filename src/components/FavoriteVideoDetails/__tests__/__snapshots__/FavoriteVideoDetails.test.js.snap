// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing FavoriteVideoDetails component should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-bdnxRM hCekCk"
      >
        <div>
          <h2
            center="true"
            class="sc-crzoAE gNzgsY title py-1"
            weight="300"
          >
            Welcome to the Challenge!
          </h2>
          <div
            class="sc-jSFjdj eUsWzQ"
          >
            <div
              class="sc-gKAaRy dBWXIC"
            >
              <div
                class="sc-gtsrHT lhUmru"
              />
              <div
                class="sc-dlnjwi CiaOg"
                data-testid="loader"
              >
                <div
                  class="sc-hKFxyN fYTRnF"
                />
              </div>
              
            </div>
          </div>
        </div>
      </div>
      <div
        class="sc-bdnxRM hCekCk"
      >
        <div>
          <div
            class="sc-iCoGMd fYbrrC"
          >
            <div
              class="sc-fujyAs bjsNot"
            >
              <iframe
                frameborder="0"
                height="600px"
                src="https://www.youtube.com/embed/undefined"
                title="video"
                width="100%"
              />
              <h2
                class="sc-crzoAE jTMVwa mb-1"
                weight="600"
              />
              <p
                class="sc-crzoAE ftDVhL"
              />
            </div>
            <div
              class="sc-fujyAs ggBxAR"
            >
              <div
                class="sc-pNWdM iLxsmF"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-bdnxRM hCekCk"
    >
      <div>
        <h2
          center="true"
          class="sc-crzoAE gNzgsY title py-1"
          weight="300"
        >
          Welcome to the Challenge!
        </h2>
        <div
          class="sc-jSFjdj eUsWzQ"
        >
          <div
            class="sc-gKAaRy dBWXIC"
          >
            <div
              class="sc-gtsrHT lhUmru"
            />
            <div
              class="sc-dlnjwi CiaOg"
              data-testid="loader"
            >
              <div
                class="sc-hKFxyN fYTRnF"
              />
            </div>
            
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-bdnxRM hCekCk"
    >
      <div>
        <div
          class="sc-iCoGMd fYbrrC"
        >
          <div
            class="sc-fujyAs bjsNot"
          >
            <iframe
              frameborder="0"
              height="600px"
              src="https://www.youtube.com/embed/undefined"
              title="video"
              width="100%"
            />
            <h2
              class="sc-crzoAE jTMVwa mb-1"
              weight="600"
            />
            <p
              class="sc-crzoAE ftDVhL"
            />
          </div>
          <div
            class="sc-fujyAs ggBxAR"
          >
            <div
              class="sc-pNWdM iLxsmF"
            />
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
